src/v/bytes/details/io_iterator_consumer.h

```diff
@@ -72,8 +72,7 @@ class io_iterator_consumer {
     template<typename Output>
     [[gnu::always_inline]] void consume_to(size_t n, Output out) {
         size_t c = consume(n, [&out](const char* src, size_t max) {
-            std::copy_n(src, max, out);
-            out += max;
+            out = std::copy_n(src, max, out);
```

Comment:
1. can you explain in the commit message why the original version of this didn't work for std::backinserter? these look identical to me.
2. add unit test

Reply:
1. edited first commit message. the gist is that std::back_inserter_iterator is a LegacyOutputIterator and it doesn't support operator+=. plus, std::copy_n already computes the new value for out
2. added iobuf_parser_consume_to unit test

Reply:
nitpick: I don't see how the assignment to `out` is observable.

Reply:
it's observable only in the outer consume_to, and it's needed in case of multiple fragments being chained

Reply:
Oh I see, it's a loop.